Use @nomiclabs/hardhat-etherscan for Etherscan verification

michalinacienciala opened the pull request:
`@nomiclabs/hardhat-etherscan` exposes verification functions for Etherscan that
are recommended by the hardhat team.

See https://hardhat.org/hardhat-runner/plugins/nomiclabs-hardhat-etherscan

Previously we were using `etherscan-verify` from `hardhat-deploy` plugin, which
had problems with verification of some contracts.

Apart from this change we also do a cleanup of supported testnets - we remove
configuration related to Ropsten and Rinkeby testnets, as [they're
going to be deprecated soon](https://blog.ethereum.org/2022/06/21/testnet-deprecation) and we've transitioned to using Goerli testnet.

michalinacienciala committed:
Use @nomiclabs/hardhat-etherscan for Etherscan verification

@nomiclabs/hardhat-etherscan exposes verification functions for Etherscan that
are recommended by the hardhat team.

See https://hardhat.org/hardhat-runner/plugins/nomiclabs-hardhat-etherscan

Previously we were using `etherscan-verify` from `hardhat-deploy` plugin, which
had problems with verification of some contracts.

Note: We introduce temporary changes helping with testing of the new
verification method. They will be removed before merge to `main`.

michalinacienciala committed:
Upgrade `hardhat-deploy` to `0.11.15` version

Previously used `0.9.29` version was causing conflicts with the
`@nomiclabs/hardhat-etherscan` plugin (`Property 'etherscan' must be of type
'EtherscanUserConfig', but here has type '{ apiKey?: string; }'` during `npm
pubish`).
According to https://github.com/wighawag/hardhat-deploy/issues/257 upgrading
`hardhat-deploy` to version `0.10.0` or higher should fix the issue.

michalinacienciala committed:
Add contract argument to `helpers.etherscan.verify`

Without it, deployment/verification was throwing following message:
```
Verifying contract 0x... on Etherscan...
Nothing to compile
NomicLabsHardhatPluginError: More than one contract was found to match the
deployed bytecode.
Please use the contract parameter with one of the following contracts:
  * contracts/governance/BaseTokenholderGovernor.sol:BaseTokenholderGovernor
  * contracts/governance/TokenholderGovernor.sol:TokenholderGovernor
  * contracts/test/TestGovernorTestSet.sol:TestTokenholderGovernor

For example:

  hardhat verify --contract contracts/Example.sol:ExampleContract <other args>

If you are running the verify subtask from within Hardhat instead:

  await run("verify:verify", {
    <other args>,
    contract: "contracts/Example.sol:ExampleContract"
  };
```

michalinacienciala committed:
Wait for 5 confirmations before veryfing TokenholderGovernor

Without the change, the deployment/verification was logging:
```
Verifying contract 0x... on Etherscan...
Nothing to compile
NomicLabsHardhatPluginError: Failed to send contract verification request.
Endpoint URL: https://api-goerli.etherscan.io/api
Reason: The Etherscan API responded that the address 0x... does not have
bytecode.
This can happen if the contract was recently deployed and this fact hasn't
propagated to the backend yet.
Try waiting for a minute before verifying your contract. If you are invoking
this from a script, try to wait for five confirmations of your contract
deployment transaction before running the verification subtask.
```

michalinacienciala committed:
Remove temporary config

michalinacienciala committed:
No longer support Ropsten and Rinkeby testnets

Both testnets are soon to be deprecated
(https://blog.ethereum.org/2022/06/21/testnet-deprecation). We've transisioned
to using Goerli testnet instead and we no longer need to have any code relating
to Ropsten/Rinkeby.

michalinacienciala commented:
Workflow dry-runned here: https://github.com/threshold-network/solidity-contracts/actions/runs/3081781328/jobs/4981728262.

cygnusv commented:
FWIW, I've had issues with v3.1.0 in this project (see commit bf90cd570ef82d56eca4b0f96352580642f563a0), so I had to pin 3.0.4.

michalinacienciala commented:
> FWIW, I've had issues with v3.1.0 in this project (see commit [bf90cd5](https://github.com/threshold-network/solidity-contracts/commit/bf90cd570ef82d56eca4b0f96352580642f563a0)), so I had to pin 3.0.4.

Do you remember what kind of issues was it, @cygnusv? Some problem with deploy/verification? I was able to deploy and verify contracts without issues...

cygnusv commented:
Thank goodness I never close my console tabs. Found it:

```
david@Davids-MacBook-Pro solidity-contracts % npx hardhat verify --network mainnet 0x57E1a87603bD1960D734243f8b2f5133911E009D 0xCdF7028ceAB81fA0C6971208e83fa7872994beE5 0x1293a54e160D1cd7075487898d65266081A15458 0xbbD3C0C794F40c4f993B03F65343aCC6fcfCb2e2 0xE47c80e8c23f6B4A1aE41c34837a0599D5D16bb0 0x1CCA7E410eE41739792eA0A24e00349Dd247680e 0x10DE37cF84202A20cae61069C617B3Aa874aF8b4
An unexpected error occurred:

TypeError: customChains is not iterable
    at getEtherscanEndpoints (/Users/david/dev/t/solidity-contracts/node_modules/@nomiclabs/hardhat-etherscan/src/network/prober.ts:33:37)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Environment._runTaskDefinition (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:219:14)
    at async Environment.run (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:131:14)
    at async SimpleTaskDefinition.verifySubtask [as action] (/Users/david/dev/t/solidity-contracts/node_modules/@nomiclabs/hardhat-etherscan/src/index.ts:209:30)
    at async Environment._runTaskDefinition (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:219:14)
    at async Environment.run (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:131:14)
    at async Environment._runTaskDefinition (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:219:14)
    at async hardhatVerify (/Users/david/dev/t/solidity-contracts/node_modules/@openzeppelin/hardhat-upgrades/src/verify-proxy.ts:110:12)
    at async OverriddenTaskDefinition._action (/Users/david/dev/t/solidity-contracts/node_modules/@openzeppelin/hardhat-upgrades/src/index.ts:155:12)
```

michalinacienciala committed:
Use fixed 3.0.4 version of @nomiclabs/hardhat-etherscan dependency

Higher version (^3.1.0) reportedly causes problems with etherscan verification
on Mainnet (`TypeError: customChains is not iterable`). See commit bf90cd570.

michalinacienciala commented:
> Thank goodness I never close my console tabs. Found it:
> 
> ```
> david@Davids-MacBook-Pro solidity-contracts % npx hardhat verify --network mainnet 0x57E1a87603bD1960D734243f8b2f5133911E009D 0xCdF7028ceAB81fA0C6971208e83fa7872994beE5 0x1293a54e160D1cd7075487898d65266081A15458 0xbbD3C0C794F40c4f993B03F65343aCC6fcfCb2e2 0xE47c80e8c23f6B4A1aE41c34837a0599D5D16bb0 0x1CCA7E410eE41739792eA0A24e00349Dd247680e 0x10DE37cF84202A20cae61069C617B3Aa874aF8b4
> An unexpected error occurred:
> 
> TypeError: customChains is not iterable
>     at getEtherscanEndpoints (/Users/david/dev/t/solidity-contracts/node_modules/@nomiclabs/hardhat-etherscan/src/network/prober.ts:33:37)
>     at processTicksAndRejections (node:internal/process/task_queues:96:5)
>     at async Environment._runTaskDefinition (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:219:14)
>     at async Environment.run (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:131:14)
>     at async SimpleTaskDefinition.verifySubtask [as action] (/Users/david/dev/t/solidity-contracts/node_modules/@nomiclabs/hardhat-etherscan/src/index.ts:209:30)
>     at async Environment._runTaskDefinition (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:219:14)
>     at async Environment.run (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:131:14)
>     at async Environment._runTaskDefinition (/Users/david/dev/t/solidity-contracts/node_modules/hardhat/src/internal/core/runtime-environment.ts:219:14)
>     at async hardhatVerify (/Users/david/dev/t/solidity-contracts/node_modules/@openzeppelin/hardhat-upgrades/src/verify-proxy.ts:110:12)
>     at async OverriddenTaskDefinition._action (/Users/david/dev/t/solidity-contracts/node_modules/@openzeppelin/hardhat-upgrades/src/index.ts:155:12)
> ```

@cygnusv, I had a bit similar-sounding issue (`TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))`) before I upgraded `hardhat-deploy` to `0.11.15`.
Just to be safe, I downgraded the `@nomiclabs/hardhat-etherscan` to `3.0.4`.

cygnusv commented on deploy/31_deploy_tokenholder_governor.ts:
```diff
@@ -23,6 +23,18 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+    await hre.ethers.provider.waitForTransaction(
+      timelock.transactionHash,
+      5,
+      300000
```
I understand the occasional need for this, but there's nothing inherent to `TokenholderGovernor` that requires an extra waiting time. It can happen to other deployments with some small probability. Can we instead add this to all Etherscan verification calls? Ideally, we can introduce it as part of a helper function.

michalinacienciala commented:
Before I do any further changes, I want to wait for the fix of the problem with failing unit tests which is introduced by upgrading `hardhat-deploy` dependency. PR for that: https://github.com/threshold-network/solidity-contracts/pull/126.

michalinacienciala committed:
Merge remote-tracking branch 'origin/main' into etherscan-verification

michalinacienciala committed:
Wait for 5 confirmations before veryfing contracts

Without the change, the verification of contracts on Etherscan could sometimes
fail with error like this:
```
Verifying contract 0x... on Etherscan...
Nothing to compile
NomicLabsHardhatPluginError: Failed to send contract verification request.
Endpoint URL: https://api-goerli.etherscan.io/api
Reason: The Etherscan API responded that the address 0x... does not have
bytecode.
This can happen if the contract was recently deployed and this fact hasn't
propagated to the backend yet.
Try waiting for a minute before verifying your contract. If you are invoking
this from a script, try to wait for five confirmations of your contract
depiloyment transaction before running the verification subtask.
```

michalinacienciala committed:
Configure Etherscan API key for use with @nomiclabs/hardhat-etherscan

In PR #126 (https://github.com/threshold-network/solidity-contracts/pull/126)
we've upgraded the `hardhat-deploy` plugin to latest version (^0.11.15). As part
of this change, the way of storing Etherscan API key has changed - to be
different from how the key is stored for @nomiclabs/hardhat-etherscan plugin
use. When we merged those changes from `main` to the current feature branch, we
no longer had a correct format of the Etherscan API key config to use it with
the `@nomiclabs/hardhat-etherscan` plugin. As we don't want to verify contracts
using `hardhat-deploy` plugin anymore, we need to adjust the config to fit
`@nomiclabs/hardhat-etherscan`'s syntax.

michalinacienciala committed:
Merge remote-tracking branch 'origin/main' into etherscan-verification

michalinacienciala committed:
Fix formatting

michalinacienciala commented:
Workflow verified in run: https://github.com/threshold-network/solidity-contracts/actions/runs/3205279132/jobs/5237683036.

michalinacienciala commented on deploy/01_deploy_t.ts:
```diff
@@ -2,14 +2,19 @@ import { HardhatRuntimeEnvironment } from "hardhat/types"
   })
 
+  if (hre.network.tags.etherscan) {
+    await hre.ethers.provider.waitForTransaction(T.transactionHash, 5, 300000)
```
I tried to use `waitForConfirmations: 1` and `waitForConfirmations: 5`, but nome of the options was good:
- `waitForConfirmations: 5` was causing hanging of the deploy function when executed on the `hardhat` network (see https://github.com/wighawag/hardhat-deploy/issues/304)
- `waitForConfirmations: 1` was not enough, the deploy of `TokenholderGovernor` was failing.
The force-pushes seen below were a result of me going back to the original solution (i.e. using `hre.ethers.provider.waitForTransaction`).
